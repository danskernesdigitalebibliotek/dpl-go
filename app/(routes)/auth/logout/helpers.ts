import { IronSession } from "iron-session"
import { NextResponse } from "next/server"
import * as client from "openid-client"

import { getUniloginClientConfig } from "@/lib/session/oauth/uniloginClient"
import {
  TSessionData,
  destroySession,
  destroySessionAndRedirectToFrontPage,
  getUniloginIdToken,
  redirectToFrontPageAndReloadSession,
} from "@/lib/session/session"

import loadAdgangsplatformenLogoutUrl from "./loadAdgangsplatformenLogoutUrl"

export const handleUniloginLogout = async (session: IronSession<TSessionData>) => {
  await logoutUniloginSSO(session)
  return destroySessionAndRedirectToFrontPage(session)
}

export const logoutUniloginSSO = async (session: IronSession<TSessionData>) => {
  const config = await getUniloginClientConfig()
  const id_token = await getUniloginIdToken()

  // If we don't have a client config, we can't do anything.
  if (!config) {
    console.error("No client config found for Unilogin.")
    return destroySessionAndRedirectToFrontPage(session)
  }

  // Call the Unilogin end session endpoint.
  // TODO: Is this where we want to redirect to if id token cannot be resolved?
  if (!id_token) {
    console.error("Could not end session in Unilogin. No id token found.")
    // Without an id token, we can't logout in the SSO.
    // So we destroy the session and redirect to the frontpage.
    return destroySessionAndRedirectToFrontPage(session)
  }
  // Resolve the end session SSO endpoint.
  const endSessionEndpoint = config.serverMetadata().end_session_endpoint
  if (!endSessionEndpoint) {
    console.error("Could not resolve Unlogin end session endpoint.")
    // We can't do anything without an end session endpoint.
    // So we destroy the session and redirect to the frontpage.
    return destroySessionAndRedirectToFrontPage(session)
  }
  // End session in Unilogin SSO.
  const endSessionUrl = client.buildEndSessionUrl(config, {
    id_token_hint: id_token,
  })
  await fetch(endSessionUrl)
}

export const handleAdgangsplatformenLogout = async (session: IronSession<TSessionData>) => {
  await destroySession(session)

  // Redirect to the logout url if available.
  const logoutUrl = await loadAdgangsplatformenLogoutUrl()
  if (logoutUrl) {
    return NextResponse.redirect(logoutUrl)
  } else {
    console.error("Could not resolve Adgangsplatformen logout url.")
    return redirectToFrontPageAndReloadSession()
  }
}
