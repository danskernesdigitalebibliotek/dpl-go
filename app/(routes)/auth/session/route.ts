import { defaultSession, getSession } from "@/lib/session/session"

export async function GET() {
  try {
    const session = await getSession()

    if (!session) {
      return Response.json({ defaultSession })
    }
    return Response.json(session)
  } catch (e) {
    return Response.json({ error: e }, { status: 500 })
  }
}
