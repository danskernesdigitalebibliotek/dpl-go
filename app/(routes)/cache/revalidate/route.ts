import { revalidatePath, revalidateTag } from "next/cache"
import { NextRequest, NextResponse } from "next/server"
import { z } from "zod"

import { getServerEnv } from "@/lib/config/env"

const paramsSchema = z.union([
  z.object({
    tags: z.string(),
    path: z.string().optional(),
    secret: z.string(),
  }),
  z.object({
    tags: z.string().optional(),
    path: z.string(),
    secret: z.string(),
  }),
])

export async function GET(request: NextRequest) {
  const secretEnv = getServerEnv("REVALIDATE_CACHE_SECRET")
  const searchParams = request.nextUrl.searchParams
  const requestParams = Object.fromEntries(searchParams.entries())

  try {
    const { tags, path, secret } = paramsSchema.parse(requestParams)

    if (!secret || secret !== secretEnv) {
      return NextResponse.json({ error: "Not authorized" }, { status: 401 })
    }

    if (path) {
      revalidatePath(path)
    }

    if (tags) {
      tags.split(",").forEach(tag => revalidateTag(tag))
    }

    return NextResponse.json({ message: "Revalidation successful" })
  } catch (e) {
    console.error("Error parsing params:", e)
    return NextResponse.json({ error: "Bad Request" }, { status: 400 })
  }
}
