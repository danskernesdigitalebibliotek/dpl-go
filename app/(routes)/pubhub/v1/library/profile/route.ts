import { NextResponse } from "next/server"

import { withAuth } from "@/app/(routes)/pubhub/(lib)/helper"

import { getLibraryProfileRequest } from "./(lib)/requests"
import { getLibraryProfileSchema } from "./(lib)/schemas"

async function getLibraryProfile() {
  const libraryProfile = getLibraryProfileSchema.transform(libraryProfileData => {
    const libraryProfileResponse = libraryProfileData.response.data
    return {
      id: libraryProfileResponse.Id,
      name: libraryProfileResponse.Name,
      ebookLoanDurationInDays: libraryProfileResponse.LoanDurationInDays,
      maxEbookAmountPerMonth: libraryProfileResponse.MaxAmountPerMonth,
      maxEbookAmountPerMonthNotificationThreshold:
        libraryProfileResponse.MaxAmountPerMonthNotificationThreshold,
      maxConcurrentEbookLoansPerBorrower: libraryProfileResponse.MaxConcurrentLoansPerBorrower,
      maxConcurrentEbookReservationsPerBorrower:
        libraryProfileResponse.MaxConcurrentReservationsPerBorrower,
      maxEbookCancellationsPerMonth: libraryProfileResponse.MaxCancellationsPerMonth,
      audioLoanDurationInDays: libraryProfileResponse.AudioLoanDurationInDays,
      maxAudioAmountPerMonth: libraryProfileResponse.MaxAudioAmountPerMonth,
      maxAudioAmountPerMonthNotificationThreshold:
        libraryProfileResponse.MaxAudioAmountPerMonthNotificationThreshold,
      maxConcurrentAudioLoansPerBorrower: libraryProfileResponse.MaxConcurrentAudioLoansPerBorrower,
      maxConcurrentAudioReservationsPerBorrower:
        libraryProfileResponse.MaxConcurrentAudioReservationsPerBorrower,
      maxAudioCancellationsPerMonth: libraryProfileResponse.MaxAudioCancellationsPerMonth,
    }
  })

  try {
    const responseData = await getLibraryProfileRequest()
    return NextResponse.json(libraryProfile.parse(responseData))
  } catch (error) {
    console.error(error)
    throw new Response("Unprocessable content", { status: 422 })
  }
}

export const GET = withAuth(getLibraryProfile)

export const dynamic = "force-dynamic"
