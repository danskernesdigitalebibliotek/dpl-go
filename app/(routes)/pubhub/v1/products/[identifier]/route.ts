import { NextRequest, NextResponse } from "next/server"

import { withAuth } from "@/app/(routes)/pubhub/(lib)/helper"
import { TUserInfo } from "@/app/(routes)/pubhub/(lib)/types"

import { getProductsRequest } from "./(lib)/requests"
import { getProductsSchema } from "./(lib)/schemas"

async function getProducts(
  request: NextRequest,
  context: { uniLoginUserInfo: TUserInfo; params: { identifier: string } }
) {
  const params = await context.params
  const identifier = params.identifier
  const getProduct = getProductsSchema.transform(productData => {
    const response = productData.response
    return {
      code: response.status.code,
      message: response.status.message,
      product: {
        title: response.data.title,
        description: response.data.description,
        costFree: response.costfree,
      },
    }
  })
  try {
    const responseData = await getProductsRequest(context.uniLoginUserInfo, identifier)
    return NextResponse.json(getProduct.parse(responseData))
  } catch (error) {
    console.error(error)
    throw new Response("Unprocessable content", { status: 422 })
  }
}

// @ts-ignore
export const GET = withAuth(getProducts)

export const dynamic = "force-dynamic"
