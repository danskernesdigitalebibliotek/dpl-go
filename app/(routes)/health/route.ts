import { promises as fs } from "fs"
import { isEmpty } from "lodash"
import { NextResponse } from "next/server"

import goConfig from "@/lib/config/goConfig"
import { loadPageData } from "@/lib/helpers/dpl-cms-content"

type TRequestsNames = "frontpage"

type THealthStatusRequestBody = {
  version: string
  requests:
    | {
        [key in TRequestsNames]: {
          status: "ok" | "error"
          message: string
        }
      }
    | Record<string, never>
}

async function getHealthStatus() {
  const requestBody: THealthStatusRequestBody = {
    version: "",
    requests: {},
  }

  try {
    // Get version from .version json file in root of project. .version file should only exist in lagoon.
    const jsonFile = await fs.readFile(".version", "utf8")
    requestBody.version = JSON.parse(jsonFile).version
  } catch (error) {
    requestBody.version = `Error reading version: ${error instanceof Error ? error.message : "Unknown error"}`
  }

  try {
    const frontpageData = await loadPageData({
      contentPath: goConfig("routes.frontpage"),
      type: "page",
    })

    if (!isEmpty(frontpageData)) {
      requestBody.requests.frontpage = {
        status: "ok",
        message: "Frontpage data loaded successfully",
      }
    } else {
      requestBody.requests.frontpage = {
        status: "error",
        message: `Frontpage data is empty`,
      }
    }
  } catch (error) {
    requestBody.requests.frontpage = {
      status: "error",
      message: `Error loading frontpage data: ${error instanceof Error ? error.message : "Unknown error"}`,
    }
  }

  return NextResponse.json(requestBody, { status: 200 })
}

export const GET = getHealthStatus

export const dynamic = "force-dynamic"
