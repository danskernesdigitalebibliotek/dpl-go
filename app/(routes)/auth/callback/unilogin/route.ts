import { NextRequest, NextResponse } from "next/server"
import * as client from "openid-client"
import type { IntrospectionResponse } from "openid-client"

import { getEnv } from "@/lib/config/env"
import goConfig from "@/lib/config/goConfig"
import { getInstitutionId, getInstitutionIds } from "@/lib/helpers/unilogin"
import { getUniloginClientConfig } from "@/lib/session/oauth/uniloginClient"
import {
  destroySession,
  getSession,
  getSessionOptions,
  setUniloginTokensOnSession,
} from "@/lib/session/session"
import { TUniloginTokenSet } from "@/lib/types/session"

import { logoutUniloginSSO } from "../../logout/helpers"
import { isUniloginUserAuthorizedToLogIn } from "./helper"
import schemas from "./schemas"

type TClaims = {
  exp: number
  iat: number
  auth_time: number
  jti: string
  iss: string
  aud: string
  sub: string
  typ: string
  azp: string
  session_state: string
  at_hash: string
  sid: string
  spec_ver: string
  has_license: string
  broker_id: string
  unilogin_loa: string
  aktoer_gruppe: string
  institution_ids: string
  loa: string
  uniid: string
}

export interface TIntrospectionResponse extends IntrospectionResponse {
  uniid: string
  institution_ids: string
}

export async function GET(request: NextRequest) {
  const session = await getSession()
  const config = await getUniloginClientConfig()
  const appUrl = getEnv("APP_URL")
  const sessionOptions = await getSessionOptions()

  if (!config || !sessionOptions) {
    return NextResponse.redirect(appUrl)
  }

  const currentSearchParams = request.nextUrl.searchParams
  const redirectUri = new URL(`${appUrl}/auth/callback/unilogin`)
  currentSearchParams.forEach((value, key) => {
    redirectUri.searchParams.append(key, value)
  })

  // TODO: When we consider the callback being stable we can remove this.
  // Maybe we can organize openid client logging in some way:
  //
  // config[client.customFetch] = async (url: string, options: RequestInit) => {
  //   // eslint-disable-next-line no-console
  //   console.log("Request URL: ", url.toString())
  //   // eslint-disable-next-line no-console
  //   console.log("Request Options: ", options)

  //   const request = new Request(url, options)
  //   return fetch(request)
  // }

  // Fetch all user/token info.
  try {
    const tokenSetResponse = await client.authorizationCodeGrant(config, redirectUri, {
      pkceCodeVerifier: session.code_verifier,
      idTokenExpected: true,
    })

    const tokenSet = schemas.tokenSet.parse(tokenSetResponse) as TUniloginTokenSet

    const introspectResponse = (await client.tokenIntrospection(
      config,
      tokenSet.access_token!
    )) as TIntrospectionResponse

    const introspect = schemas.introspect.parse(introspectResponse)

    const claims = tokenSetResponse.claims()! as TClaims

    // UserInfo Request
    const userInfoResponse = await client.fetchUserInfo(config, tokenSet.access_token, claims.sub)

    const userinfo = schemas.uniLoginUserInfo.parse(userInfoResponse)

    // Set basic session info.
    session.isLoggedIn = true
    session.type = "unilogin"

    // Set token info.
    setUniloginTokensOnSession(session, tokenSet)

    const institutionId = getInstitutionId(introspect.institution_ids)
    // Check if user is authorized to log.
    const isAuthorized = await isUniloginUserAuthorizedToLogIn(institutionId, claims)
    if (isAuthorized === false) {
      // Make sure that the user is logged out remotely first. And destroy session.
      await logoutUniloginSSO(session)
      await destroySession(session)
      // Redirect user to login not authorized page.
      return NextResponse.redirect(
        `${getEnv("APP_URL")}/${goConfig("routes.login-not-authorized")}`
      )
    }

    // Set user info.
    // TODO: After Publizon allows DDF test users to loan, we can remove thie if statement.
    session.uniLoginUserInfo = {
      sub: userinfo.sub,
      uniid: introspect.uniid,
      // TODO: Rename this into institutionIds
      institution_ids:
        institutionId === "A04441" ? ["101047"] : getInstitutionIds(introspect.institution_ids),
    }
    session.user = {
      // Unilogin does not provide a name, so we set it to undefined.
      name: undefined,
      username: introspect.uniid,
    }

    await session.save()
  } catch (error) {
    console.error(error)
    // Make sure that the user is logged out remotely first. And destroy session.
    await logoutUniloginSSO(session)
    await destroySession(session)
    return NextResponse.redirect(`${getEnv("APP_URL")}/${goConfig("routes.login-failed-unilogin")}`)
  }

  return NextResponse.redirect(`${getEnv("APP_URL")}/user/profile`)
}
