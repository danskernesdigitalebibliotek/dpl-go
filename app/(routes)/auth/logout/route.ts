import { destroySessionAndRedirectToFrontPage, getSession } from "@/lib/session/session"

import { handleAdgangsplatformenLogout, handleUniloginLogout } from "./helpers"

export async function GET() {
  const session = await getSession()

  if (session.type === "unilogin") {
    return handleUniloginLogout(session)
  }

  if (session.type === "adgangsplatformen") {
    return handleAdgangsplatformenLogout(session)
  }

  return destroySessionAndRedirectToFrontPage(session)
}
