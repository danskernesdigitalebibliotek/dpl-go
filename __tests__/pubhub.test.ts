import { testApiHandler } from "next-test-api-route-handler"
import { describe, expect, it } from "vitest"

import { getV1UserLoansSoapData } from "@/__tests__/mocks/pubhub"
import * as apiEndpoint from "@/app/pubhub/v1/user/loans/route"
import * as sessionFunctions from "@/lib/session/session"
import * as clientFunctions from "@/lib/soap/publizon/v2_7/generated/getlibraryuserorderlist/client"

import { testSilently } from "./helpers"

describe("Pubhub local API", () => {
  testSilently("Returns unauthorized for anonymous user at GET /v1/user/loans", async () => {
    await testApiHandler({
      // @ts-ignore
      appHandler: apiEndpoint,
      url: "/pubhub/v1/user/loans",
      async test({ fetch }) {
        const res = await fetch({ method: "GET" })
        expect(res.status).equal(401)
      },
    })
  })

  it("Returns authorized for logged in users at GET /v1/user/loans", async () => {
    vi.spyOn(sessionFunctions, "getSession").mockResolvedValue(
      // @ts-ignore
      Promise.resolve({
        isLoggedIn: true,
        type: "unilogin",
        uniLoginUserInfo: { uniid: "100006cbab", institution_ids: ["A04441"] },
      })
    )
    vi.spyOn(clientFunctions, "createClientAsync").mockResolvedValue(
      // @ts-ignore
      Promise.resolve({
        GetLibraryUserOrderListAsync: async () => {
          return Promise.resolve([getV1UserLoansSoapData])
        },
      })
    )
    await testApiHandler({
      // @ts-ignore
      appHandler: apiEndpoint,
      url: "/pubhub/v1/user/loans",
      async test({ fetch }) {
        const res = await fetch({ method: "GET" })
        expect(res.status).equal(200)
      },
    })
  })

  it("Returns same output from local & external GET /v1/user/loans", async () => {
    vi.spyOn(sessionFunctions, "getSession").mockResolvedValue(
      // @ts-ignore
      Promise.resolve({
        isLoggedIn: true,
        type: "unilogin",
        uniLoginUserInfo: { uniid: "100006cbab", institution_ids: ["A04441"] },
      })
    )
    vi.spyOn(clientFunctions, "createClientAsync").mockResolvedValue(
      // @ts-ignore
      Promise.resolve({
        GetLibraryUserOrderListAsync: async () => {
          return Promise.resolve([getV1UserLoansSoapData])
        },
      })
    )
    await testApiHandler({
      // @ts-ignore
      appHandler: apiEndpoint,
      url: "/pubhub/v1/user/loans",
      async test({ fetch }) {
        const res = await fetch({ method: "GET" })
        const data = await res.json()
        expect(data).toMatchSnapshot()
      },
    })
  })
})
