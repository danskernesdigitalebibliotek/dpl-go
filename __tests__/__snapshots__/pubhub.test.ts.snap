// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pubhub local API > Returns same output from local & external GET /v1/user/loans 1`] = `
{
  "libraryData": {
    "maxAmountPerMonth": 3,
    "maxConcurrentAudiobookLoansPerBorrower": 0,
    "maxConcurrentEbookLoansPerBorrower": 3,
  },
  "loans": [
    {
      "libraryBook": {
        "identifier": "9788772144467",
      },
      "loanExpireDateUtc": "2025-04-10T14:28:57",
      "orderDateUtc": "2025-03-11T13:30:04",
      "orderId": "7cab0a54-2a22-4172-862c-b3ed346e1da0",
    },
  ],
  "userData": {
    "audiobookLoansRemaining": 0,
    "ebookLoansRemaining": 2,
    "totalAudioLoans": 0,
    "totalEbookLoans": 1,
    "totalLoans": 1,
  },
}
`;
